Pre-registering certain standard classes to better support long term persistence scenarios

gdiet opened the pull request:
This PR is about registering standard scala classes that are likely to be used when long term persisting data, for example in event stores:

When using kryo/chill for long term storage, it is advisable to run Kryo with setRegistrationRequired(true). However, a number of standard scala classes that are useful in such scenarios are not pre-registered in chill, most of them located in scala.collections.immutable. This has puzzled others as well, see for example the chill issue #227 which is fixed by this PR.

Some of the classes I have added to the standard registration in this pull request:

scala.None
scala.collection.immutable.::
scala.collection.immutable.Nil
scala.collection.immutable.ListSet.EmptyListSet
scala.collection.immutable.Range
scala.collection.mutable.WrappedArray.ofRef
int[]

To better support the long term persistence scenario, I have added the RegistrationIdsSpec, which will fail if the pre-registered IDs change, because changing them should be avoided if possible to ensure backwards compatibility. Or, if changing pre-registered IDs is really inevitable, a compatibility note should be added to the release notes.

If this PR is accepted, I plan to complement it with two more PRs: One PR for the chill documentation with notes on how kryo/chill can be used for long term persisting events (lessons learned). The other PR would be another spec ensuring that the binary representation of serialized classes is not changed unintendedly.

gdiet committed:
Update to latest scala versions

gdiet committed:
Fix for deprecation warning

gdiet committed:
Fixed typo in spec and removed unused import

gdiet committed:
Added test to ensure long term stability of registration IDs

gdiet committed:
Renamed spec

gdiet committed:
Require registration for empty set

gdiet committed:
Registration for empty set

gdiet committed:
Require registration for hash sets

gdiet committed:
Registration for hash sets

gdiet committed:
Require registration for list set

gdiet committed:
Registration for list set

gdiet committed:
Require registration for list, queue, range

gdiet committed:
Registration for list, queue, range

gdiet committed:
Require registration for hash maps, list maps, None, and arrays

gdiet committed:
Registration for hash maps, list maps, None, and arrays

gdiet committed:
Require registration for wrapped arrays

gdiet committed:
Registration for wrapped arrays

gdiet committed:
Require registration for wrapped strings

gdiet committed:
Registration for wrapped strings

gdiet committed:
Registration for streams

gdiet committed:
Registration for TreeSet and TreeMap

gdiet committed:
Added missing registrations to test

gdiet committed:
Registration for the empty map

gdiet committed:
Added missing registrations to test

gdiet committed:
found a problem when serializing streams

gdiet committed:
fixed stream serializer

gdiet committed:
finished test examples

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/chill?pullRequest=302) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/twitter/chill?pullRequest=302) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: gdiet<br/>:x: Georg Dietrich<br/><hr/>**Georg Dietrich** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/chill?pullRequest=302) it.</sub>

gdiet commented:
Yes, Travis CI build failed, but :) this is because travis uses scala 2.10.6, and with scala 2.10.7 the tests run green (at least locally). So I propose that the Travis CI is reconfigured to use scala 2.10.7.

johnynek commented:
@gdiet can you bump to modern scalas here:

https://github.com/twitter/chill/blob/develop/.travis.yml

then you should be green?

johnynek commented:
Thanks for this PR, by the way!

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ScalaKryoInstantiator.scala:
```diff
@@ -121,15 +98,32 @@ class ScalaCollectionsRegistrar extends IKryoRegistrar {
      */
     newK
       // wrapper array is abstract
+      .registerClasses(Seq(
```
can you add all new items at the end of the list? Otherwise, I think we could change the tokens and some people have (ill-advisedly) persisted kryo data into long term storage.

gdiet replied:
yes... I will follow your next proposal and put all new items into a separate registrar.

johnynek commented on chill-scala/src/test/scala/com/twitter/chill/RegistrationIdsSpec.scala:
```diff
@@ -0,0 +1,179 @@
+
+class RegistrationIdsSpec extends WordSpec with Matchers {
+  """
+    |Projects using chill to long term persist serialized data (for example in event
```
so this test protects the future, but not people in the past right? They all get broken right?

Could we work around that? Maybe add a new registrar that people can opt in to adding?

gdiet replied:
Right, this would break compatibility for projects that have already used chill for long term storage. Defining a separate registrar is OK for me (although the order of entries in the registrar will be less orderly that way).

I'm thinking along these lines currently:
* Rename AllScalaRegistrar -> AllScalaRegistrar_0_9_2 for those who need the old registrations.
* Create a new AllScalaRegistrar that first registers everything from AllScalaRegistrar_0_9_2, then registers all the new items from this PR.

Does this sound OK?

johnynek reviewed (commented):
This looks like a great improvement. I appreciate the tests. Thank you.

gdiet committed:
bumping up scala versions for travis ci

Commit:
revert ScalaKryoInstantiator to redesign changes in backward compatible way

Commit:
introduced 0.9.2 compatibility registrar

Commit:
on the way towards getting things to work again

com.twitter.chill.StandardDataRegistrationsSpec still fails, needs investigation

Commit:
fixed wrapped array serialization and re-ordered additional registrations in AllScalaRegistrar

Commit:
added explaining scaladoc comment

gdiet committed:
improved scaladoc comments for changed and new stuff
